File: src/main/java/komu/blunt/core/CoreConstantExpression.java
package komu.blunt.core;

import komu.blunt.asm.Assembler;
import komu.blunt.asm.Instructions;
import komu.blunt.asm.Linkage;
import komu.blunt.asm.Register;

import static com.google.common.base.Preconditions.checkNotNull;

public final class CoreConstantExpression extends CoreExpression {

    private final Object value;

    public CoreConstantExpression(Object value) {
        this.value = checkNotNull(value);
    }

    @Override
    public void assemble(Assembler asm, Instructions instructions, Register target, Linkage linkage) {
        instructions.loadConstant(target, value);
        instructions.finishWithLinkage(linkage);
    }

    @Override
    public String toString() {
        return String.valueOf(value);
    }
}

File: src/main/kotlin/komu/blunt/core/CoreConstantExpression.kt
package komu.blunt.core

import komu.blunt.asm.*

class CoreConstantExpression(val value: Any) : CoreExpression() {

    override fun assemble(asm: Assembler, target: Register, linkage: Linkage) =
        instructions {
            loadConstant(target, value)
            finishWithLinkage(linkage)
        }

    override fun simplify() = this
    override fun toString() = value.toString()
}

